### PR TITLE
Closes #66. Add CURLOPT_FILE option support

### LINKED ISSUE #66: How to download a file from the remote server?
The curl have an option CURLOPT_FILE.
Option is usable for download file without of memory usage trouble.
Will you implement something like this?

### PR DESCRIPTION
**Pull request by alexkart**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #66 

This feature is quite simple to implement. I'll also add tests later.
Work still in progress and I need an opinion about the interface. How to name the method? I've just named it `setFile()` for now.
```php
$fh = fopen('path/to/file.png', 'w');
$request->setMethod('GET')
            ->setUrl('http://path/to/file.png')
            ->setFile($fh);
```
Also, do we need to accept only `$fh` of resource type or string as well and if string (path to file) is passed instead of file handler open a file with this path?

### CODE REVIEW

**Comment by samdark** [comment]
What if you want to upload multiple files?

File: src/Request.php [hunk 1]
```diff
@@ -494,4 +499,20 @@ private function getFormatter()
...
+     */
+    public function getFile()
+    {
+        return $this->_file;
+    }
+
+    /**
+     * @param resource $file
+     */
+    public function setFile($file)
```
Comment (samdark) [inline src/Request.php]: It's easy to confuse it with a file meant for uploading it. I'd name it something like `setOutputFile()`.

File: src/Request.php [hunk 2]
```diff
@@ -494,4 +499,20 @@ private function getFormatter()
     {
         return $this->client->getFormatter($this->getFormat());
     }
+
+    /**
+     * @return resource
+     */
+    public function getFile()
```
Comment (samdark) [inline src/Request.php]: Do we need a getter?

Comment (alexkart) [reply src/Request.php]: Why not? But I can remove it, if you recommend so.

Comment (samdark) [reply src/Request.php]: It's better not to add it if we don't have any use case for it.

Comment (alexkart) [reply src/Request.php]: Actually, we use it in CurlTransport class.

Comment (samdark) [reply src/Request.php]: Right. Never mind then.

**Comment by alexkart** [comment]
Ready for review.

File: docs/guide/basic-usage.md [hunk 3]
```diff
@@ -230,3 +230,23 @@ $client->post('account/profile', ['birthDate' => '10/11/1982'])
     ->setCookies($loginResponse->cookies) // transfer response cookies to request
     ->send();
 ```
+
+## Downloading files using `CurlTransport`
+
+You can download a file without loading entire file content into memory (it is especially useful for big files). 
+Use `setOutputFile()` method to pass a stream resource (using `fopen()`, for example) to the request object. It will be 
+passed to the `CURLOPT_FILE` CURL option.
```
Comment (machour) [inline docs/guide/basic-usage.md]:
```suggestion
passed to the `CURLOPT_FILE` cURL option.
```

File: docs/guide/basic-usage.md [hunk 4]
```diff
@@ -230,3 +230,23 @@ $client->post('account/profile', ['birthDate' => '10/11/1982'])
     ->setCookies($loginResponse->cookies) // transfer response cookies to request
     ->send();
 ```
+
+## Downloading files using `CurlTransport`
+
+You can download a file without loading entire file content into memory (it is especially useful for big files).
```
Comment (machour) [inline docs/guide/basic-usage.md]:
```suggestion
You can download a file without loading its entire content into memory (it is especially useful for big files). 
```

File: src/Request.php [hunk 5]
```diff
@@ -494,4 +499,20 @@ private function getFormatter()
     {
         return $this->client->getFormatter($this->getFormat());
     }
+
+    /**
+     * @return resource
```
Comment (samdark) [inline src/Request.php]: Need `@since 2.0.9` and description.

File: src/Request.php [hunk 6]
```diff
@@ -494,4 +499,20 @@ private function getFormatter()
...
+     */
+    public function getOutputFile()
+    {
+        return $this->_outputFile;
+    }
+
+    /**
+     * @param resource $file
```
Comment (samdark) [inline src/Request.php]: Need `@since 2.0.9` and description.

**Comment by samdark** [comment]
Merged. Thank you!

**Comment by Thoulah** [comment]
I can't get this to work:
```
	public static function saveUrl(string $url, string $file): Response {
var_dump($url);
var_dump($file);
		$fh = fopen($file, 'w');
var_dump($fh);
		$client = new Client(['transport' => CurlTransport::class]);
var_dump($client);

		$response = $client->createRequest()
			->setMethod('GET')
			->setUrl($url)
			->setOutputFile($fh)
			->send();
		fclose($fh);
		return $response;
	}
```

Resulting dumps:
/htdocs/yii/mr42/mister42/../../../.cache/yii/mister42/oui.csv
```
string(41) "http://standards-oui.ieee.org/oui/oui.csv"
string(62) "/htdocs/yii/mr42/mister42/../../../.cache/yii/mister42/oui.csv"
resource(68) of type (stream)
object(yii\httpclient\Client)#32 (10) {
  ["baseUrl"]=>
  NULL
  ["formatters"]=>
  array(0) {
  }
  ["parsers"]=>
  array(0) {
  }
  ["requestConfig"]=>
  array(0) {
  }
  ["responseConfig"]=>
  array(0) {
  }
  ["contentLoggingMaxSize"]=>
  int(2000)
  ["_transport":"yii\httpclient\Client":private]=>
  string(28) "yii\httpclient\CurlTransport"
  ["_events":"yii\base\Component":private]=>
  array(0) {
  }
  ["_eventWildcards":"yii\base\Component":private]=>
  array(0) {
  }
  ["_behaviors":"yii\base\Component":private]=>
  NULL
}
```
Looks good to me. But...:

```
Exception 'Error' with message 'Call to a member function send() on null'

in /htdocs/yii/mr42/mister42/models/Webrequest.php:53

Stack trace:
#0 /htdocs/yii/mr42/mister42/commands/FeedController.php(75): app\models\Webrequest::saveUrl('http://standard...', '/htdocs/yii/mr4...')
#1 [internal function]: app\commands\FeedController->actionOui()
#2 /htdocs/yii/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#3 /htdocs/yii/vendor/yiisoft/yii2/base/Controller.php(157): yii\base\InlineAction->runWithParams(Array)
#4 /htdocs/yii/vendor/yiisoft/yii2/console/Controller.php(148): yii\base\Controller->runAction('oui', Array)
#5 /htdocs/yii/vendor/yiisoft/yii2/base/Module.php(528): yii\console\Controller->runAction('oui', Array)
#6 /htdocs/yii/vendor/yiisoft/yii2/console/Application.php(180): yii\base\Module->runAction('feed/oui', Array)
#7 /htdocs/yii/vendor/yiisoft/yii2/console/Application.php(147): yii\console\Application->runAction('feed/oui', Array)
#8 /htdocs/yii/vendor/yiisoft/yii2/base/Application.php(386): yii\console\Application->handleRequest(Object(yii\console\Request))
#9 /htdocs/yii/mr42/App.php(23): yii\base\Application->run()
#10 /htdocs/yii/mr42/yii(4): App->__construct(true)
#11 {main}
```

**Comment by machour** [comment]
@Thoulah Thank you for catching this, will be fixed by #184

**Comment by Thoulah** [comment]
On a sidenote, it would look nicer if we can just gave it a filepath instead of a handle. Then internally the handle can be opened, written and closed?

**Comment by alexkart** [comment]
@Thoulah, I thought about it, but I'm not sure if it is a responsibility of the http client to open files, and it can be not only a file handle but any stream handle where client can write to. It basically the same why CURL library doesn't do that and requires resourse and not filepath.

**Comment by Thoulah** [comment]
Fair enough :)

**Pull request merged**